File: AlgorithmLcr/Try3-Component types with known topology and corresponding positions/topofit_id/nodal.py
"""Batched nodal analysis with adjoint sensitivities (DESIGN.md sec.4).

Stamping (Try2 netgraph_id/nodal.py convention): edge (u,v) with
admittance y contributes Y[u,u]+=y, Y[v,v]+=y, Y[u,v]-=y, Y[v,u]-=y.
Node 0 is grounded; with unit current injected at node 1,

    Z(s) = (Y_red^{-1})[0, 0]

over the (V-1)x(V-1) reduced matrix of non-ground nodes (node 1 first).

Parameter vector theta = log10 of linear values, laid out per edge in
order: R -> [log R];  C -> [log C];  L -> [log L, log Rd].

The Jacobian dZ/dtheta comes from the adjoint (Tellegen/direct) method:
with x = Y_red^{-1} e_0 (already computed for Z),

    dZ/dtheta_t = -x^T (dY/dtheta_t) x.

dY/dtheta_t has support only on edge(t)'s 2x2 stamp block, so

    dZ/dtheta_t[k] = -(dy_t/dtheta_t)[k] * (x[k, ri] - x[k, rj])^2

with x[ground] := 0.  After one batched LU per frequency the whole p x M
Jacobian costs O(M*p) -- essentially free (Director & Rohrer 1969).
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from .graph import PortOpenError

LN10 = float(np.log(10.0))
# sanitized stand-in for non-finite Z (singular Y at an exact internal
# resonance); large enough to dominate any relative residual
_BIG_Z = 1e12


@dataclass
class NodalModel:
    """Reduced-matrix evaluator for an edge list [(u, v, kind), ...]."""

    edges: list              # [(u, v, kind)]
    node_of: list            # reduced index -> original label; [0] is node 1
    n_params: int
    edge_param_slices: list  # per edge, (start, count); L: (start, 2)
    ri_list: list            # per edge reduced index of u (None = ground)
    rj_list: list            # per edge reduced index of v (None = ground)

    @classmethod
    def from_edges(cls, edges: list) -> "NodalModel":
        nodes = {n for (u, v, _k) in edges for n in (u, v)}
        for t in (0, 1):
            if t not in nodes:
                raise PortOpenError("port node {} not incident to any edge".format(t))
        others = sorted(nodes - {0, 1})
        node_of = [1] + others
        index_of = {lab: i for i, lab in enumerate(node_of)}

        n_params = 0
        slices = []
        ri_list, rj_list = [], []
        for (u, v, kind) in edges:
            cnt = 2 if kind == "L" else 1
            slices.append((n_params, cnt))
            n_params += cnt
            ri_list.append(index_of.get(u))
            rj_list.append(index_of.get(v))
        return cls(edges=list(edges), node_of=node_of, n_params=n_params,
                   edge_param_slices=slices, ri_list=ri_list, rj_list=rj_list)

    @property
    def n_nodes_red(self) -> int:
        return len(self.node_of)

    # -- admittances ---------------------------------------------------------
    def _edge_y_dy(self, kind: str, vals: np.ndarray, s: np.ndarray):
        """y(s) and dy/dlog10param for one edge; vals = linear values.

        R: y=1/r,        dy/dt = -ln10 * y
        C: y=s*c,        dy/dt = +ln10 * y
        L: y=1/(rd+s*l), dy/dl = -ln10*l*s*y^2, dy/drd = -ln10*rd*y^2
        """
        if kind == "R":
            y = 1.0 / vals[0]
            dy = (-LN10 * y) * np.ones_like(s, dtype=complex)
            return y * np.ones_like(s, dtype=complex), [dy]
        if kind == "C":
            y = s * vals[0]
            return y, [(LN10 * y)]
        l, rd = vals
        z = rd + s * l
        y = 1.0 / z
        y2 = y * y
        return y, [(-LN10 * l * s) * y2, (-LN10 * rd) * y2]

    def _solve(self, Y: np.ndarray) -> np.ndarray:
        """Solve Y x = e_0 batched over frequencies; returns x (M, k)."""
        M, k, _ = Y.shape
        rhs = np.zeros((M, k, 1), dtype=complex)
        rhs[:, 0, 0] = 1.0
        try:
            return np.linalg.solve(Y, rhs)[:, :, 0]
        except np.linalg.LinAlgError:
            x = np.full((M, k), np.nan, dtype=complex)
            for m in range(M):
                try:
                    x[m] = np.linalg.lstsq(Y[m], rhs[m, :, 0], rcond=None)[0]
                except np.linalg.LinAlgError:
                    pass
            return x

    def z_and_jac(self, theta: np.ndarray, s: np.ndarray):
        """Z (M,) and full Jacobian dZ/dtheta (p, M)."""
        theta = np.asarray(theta, dtype=float)
        s = np.asarray(s, dtype=complex)
        M = len(s)
        k = self.n_nodes_red
        vals = np.power(10.0, theta)

        Y = np.zeros((M, k, k), dtype=complex)
        edge_dy = []  # per edge: list of dy arrays (len 1 or 2)
        for e, (u, v, kind) in enumerate(self.edges):
            sl, cnt = self.edge_param_slices[e]
            y, dys = self._edge_y_dy(kind, vals[sl:sl + cnt], s)
            edge_dy.append(dys)
            ri, rj = self.ri_list[e], self.rj_list[e]
            if ri is not None:
                Y[:, ri, ri] += y
            if rj is not None:
                Y[:, rj, rj] += y
            if ri is not None and rj is not None:
                Y[:, ri, rj] -= y
                Y[:, rj, ri] -= y

        x = self._solve(Y)
        Z = x[:, 0].copy()
        bad = ~np.isfinite(Z) | (np.abs(Z) > _BIG_Z)
        if bad.any():
            Z[bad] = _BIG_Z

        J = np.zeros((self.n_params, M), dtype=complex)
        for e, (u, v, kind) in enumerate(self.edges):
            ri, rj = self.ri_list[e], self.rj_list[e]
            d = np.zeros(M, dtype=complex)
            if ri is not None:
                d = d + x[:, ri]
            if rj is not None:
                d = d - x[:, rj]
            d2 = d * d
            sl, cnt = self.edge_param_slices[e]
            for j, dy in enumerate(edge_dy[e]):
                dz = -(dy * d2)
                dz[bad] = 0.0
                dz[~np.isfinite(dz)] = 0.0
                J[sl + j] = dz
        return Z, J

    def z_linear(self, vals: np.ndarray, s: np.ndarray) -> np.ndarray:
        """Z from linear values (synthetic truth, reports)."""
        vals = np.asarray(vals, dtype=float)
        if np.any(vals <= 0):
            raise ValueError("linear values must be positive")
        Z, _ = self.z_and_jac(np.log10(vals), s)
        return Z

    def elasticity(self, theta: np.ndarray, s: np.ndarray) -> np.ndarray:
        """E[t, k] = dlnZ_k / dln(value_t) at theta (p, M).

        dZ/dv_t = J_t / (ln10 * v_t)  so  E_t = (v_t/Z) dZ/dv_t = J_t/(Z ln10)
        -- the value factors cancel (regression-tested against central
        differences in tests/test_nodal.py)."""
        theta = np.asarray(theta, dtype=float)
        Z, J = self.z_and_jac(theta, s)
        return J / (Z[None, :] * LN10)


def model_from_reduced(red) -> NodalModel:
    """NodalModel over a ReductionResult's reduced edges."""
    return NodalModel.from_edges([(e.u, e.v, e.kind) for e in red.edges])

File: AlgorithmLcr/Try3-Component types with known topology and corresponding positions/cppversion/src/nodal.cpp
#include "nodal.hpp"

#include <algorithm>
#include <cmath>
#include <map>
#include <set>

namespace tf {

NodalModel NodalModel::fromEdges(const std::vector<std::tuple<int, int, char>>& edges) {
    std::set<int> nodes;
    for (const auto& [u, v, k] : edges) {
        (void)k;
        nodes.insert(u);
        nodes.insert(v);
    }
    for (int t = 0; t < 2; ++t)
        if (!nodes.count(t))
            throw PortOpenError("port node " + std::to_string(t) +
                                " not incident to any edge");
    std::vector<int> others;
    for (int n : nodes)
        if (n != 0 && n != 1) others.push_back(n);
    std::sort(others.begin(), others.end());
    std::vector<int> nodeOf{1};
    nodeOf.insert(nodeOf.end(), others.begin(), others.end());
    std::map<int, int> indexOf;
    for (size_t i = 0; i < nodeOf.size(); ++i) indexOf[nodeOf[i]] = (int)i;

    NodalModel m;
    m.edges = edges;
    m.nodeOf = nodeOf;
    for (const auto& [u, v, kind] : edges) {
        int cnt = kind == 'L' ? 2 : 1;
        m.edgeParamSlices.push_back({m.nParams, cnt});
        m.nParams += cnt;
        auto itU = indexOf.find(u);
        auto itV = indexOf.find(v);
        m.riList.push_back(itU == indexOf.end() ? -1 : itU->second);
        m.rjList.push_back(itV == indexOf.end() ? -1 : itV->second);
    }
    return m;
}

namespace {

// y(s) and dy/dlog10param for one edge at one frequency (vals = 10^theta).
inline void edgeYdY(char kind, double v1, double v2, Complex s, Complex& y,
                    Complex& dy1, Complex& dy2) {
    if (kind == 'R') {
        y = Complex(1.0 / v1, 0.0);
        dy1 = -kLn10 * y;
        dy2 = Complex(0.0, 0.0);
        return;
    }
    if (kind == 'C') {
        y = s * v1;
        dy1 = kLn10 * y;
        dy2 = Complex(0.0, 0.0);
        return;
    }
    Complex z(v2, 0.0);
    z += s * v1;
    y = 1.0 / z;
    Complex y2 = y * y;
    dy1 = (-kLn10 * v1 * s) * y2;
    dy2 = Complex(-kLn10 * v2, 0.0) * y2;
}

}  // namespace

void NodalModel::zAndJac(const std::vector<double>& theta,
                         const std::vector<Complex>& s, std::vector<Complex>& Z,
                         std::vector<Complex>& J) const {
    const int M = (int)s.size();
    const int k = nNodesRed();
    const int p = nParams;
    std::vector<double> vals(p);
    for (int t = 0; t < p; ++t) vals[t] = std::pow(10.0, theta[(size_t)t]);

    Z.assign((size_t)M, Complex(0.0, 0.0));
    J.assign((size_t)p * M, Complex(0.0, 0.0));

    for (int m = 0; m < M; ++m) {
        std::vector<Complex> Y((size_t)k * k, Complex(0.0, 0.0));
        // per-edge admittances and log-derivatives at this frequency
        std::vector<Complex> yE(edges.size()), dE1(edges.size()), dE2(edges.size());
        for (size_t e = 0; e < edges.size(); ++e) {
            auto [sl, cnt] = edgeParamSlices[e];
            edgeYdY(std::get<2>(edges[e]), vals[(size_t)sl],
                    cnt == 2 ? vals[(size_t)sl + 1] : 0.0, s[(size_t)m], yE[e], dE1[e],
                    dE2[e]);
            int ri = riList[e], rj = rjList[e];
            if (ri >= 0) Y[(size_t)ri * k + ri] += yE[e];
            if (rj >= 0) Y[(size_t)rj * k + rj] += yE[e];
            if (ri >= 0 && rj >= 0) {
                Y[(size_t)ri * k + rj] -= yE[e];
                Y[(size_t)rj * k + ri] -= yE[e];
            }
        }
        std::vector<Complex> x((size_t)k, Complex(0.0, 0.0));
        std::vector<Complex> b((size_t)k, Complex(0.0, 0.0));
        b[0] = Complex(1.0, 0.0);
        std::vector<Complex> A = Y;
        if (!luSolveComplex(A, k, b)) {
            // singular: sanitized big-Z, Jacobian rows zeroed (measure-zero)
            Z[(size_t)m] = Complex(kBigZ, 0.0);
            continue;
        }
        x = b;
        Complex z0 = x[0];
        if (!std::isfinite(z0.real()) || !std::isfinite(z0.imag()) ||
            std::abs(z0) > kBigZ)
            z0 = Complex(kBigZ, 0.0);
        Z[(size_t)m] = z0;
        for (size_t e = 0; e < edges.size(); ++e) {
            Complex d(0.0, 0.0);
            if (riList[e] >= 0) d += x[(size_t)riList[e]];
            if (rjList[e] >= 0) d -= x[(size_t)rjList[e]];
            Complex d2 = d * d;
            auto [sl, cnt] = edgeParamSlices[e];
            Complex dys[2] = {dE1[e], dE2[e]};
            for (int j = 0; j < cnt; ++j) {
                Complex dz = -(dys[j] * d2);
                if (!std::isfinite(dz.real()) || !std::isfinite(dz.imag()) ||
                    std::abs(z0) >= kBigZ)
                    dz = Complex(0.0, 0.0);
                J[(size_t)(sl + j) * M + (size_t)m] = dz;
            }
        }
    }
}

std::vector<Complex> NodalModel::zLinear(const std::vector<double>& vals,
                                         const std::vector<Complex>& s) const {
    for (double v : vals)
        if (!(v > 0.0)) throw std::invalid_argument("linear values must be positive");
    std::vector<double> theta(vals.size());
    for (size_t i = 0; i < vals.size(); ++i) theta[i] = std::log10(vals[i]);
    std::vector<Complex> Z, J;
    zAndJac(theta, s, Z, J);
    return Z;
}

void NodalModel::elasticity(const std::vector<double>& theta,
                            const std::vector<Complex>& s,
                            std::vector<Complex>& E) const {
    std::vector<Complex> Z, J;
    zAndJac(theta, s, Z, J);
    const int M = (int)s.size();
    E.assign(J.size(), Complex(0.0, 0.0));
    for (int t = 0; t < nParams; ++t)
        for (int m = 0; m < M; ++m)
            E[(size_t)t * M + (size_t)m] =
                J[(size_t)t * M + (size_t)m] / (Z[(size_t)m] * kLn10);
}

NodalModel modelFromReduced(const ReductionResult& red) {
    std::vector<std::tuple<int, int, char>> edges;
    edges.reserve(red.edges.size());
    for (const auto& e : red.edges) edges.push_back({e.u, e.v, e.kind});
    return NodalModel::fromEdges(edges);
}

}  // namespace tf
